File: stackMachine.py
#add modes

MAXVAL = 1000000000000

class StackMachine:
    def __init__(self, maxjumps, maxbits, stackSize=10, defaultVal=1):
        self.maxbits = maxbits
        self.maxjumps = maxjumps
        self.stackSize = stackSize
        self.defaultVal = defaultVal
        self.suffix = [self._out, self._push(0), self._jmp]

        self.opcodes = []
        self.fillOps()

    def reset(self):
        self.curjumps = 0
        self.cmp = False
        self.stack = [self.defaultVal] * self.stackSize
        self.output = []
        self.stackptr = 0
        self.iptr = 0
        self.running = True
        
    def execute(self, prog):
        self.reset()
        prog.extend(self.suffix)
        while self.iptr < len(prog):
            if not prog[self.iptr]():
                self.iptr += 1
            
        return self.output

    def pop(self):
        x = self.stack[self.stackptr]
        self.stackptr = (self.stackptr - 1) % self.stackSize
        return x

    def peek(self):
        return self.stack[self.stackptr]

    def push(self, x):
        if x > MAXVAL or x < -MAXVAL:
            x %= MAXVAL
        self.stackptr = (self.stackptr + 1) % self.stackSize
        self.stack[self.stackptr] = x

    def _jmp(self):
        self.curjumps += 1
        if self.curjumps >= self.maxjumps:
            return False

        self.iptr = self.pop() % self.stackSize
        return True

    def _out(self):
        self.output.append(int(self.peek()))

    def _push(self, x):
        def f():
            self.push(x)

        return f
    
    def _swap(self):
        a, b = self.pop(), self.pop()
        self.push(b)
        self.push(a)

    def _rot(self):
        self.pop()

    def _over(self):
        pass

    def _dup(self):
        self.push(self.peek())

    def _iftrue(self):
        if not self.cmp:
            self.iptr += 1

    def _less(self):
        self.cmp = (self.pop() < self.pop())

    def _equal(self):
        self.cmp = (self.pop() == self.pop())

    def _plus(self):
        self.push(self.pop() + self.pop())

    def _neg(self):
        self.push(-self.pop())

    def _mul(self):
        self.push(self.pop() * self.pop())

    def _div(self):
        a = self.pop()
        b = self.pop()
        if b == 0:
            if a >= 0:
                self.push(MAXVAL)
            else:
                self.push(-MAXVAL)
        else:
            self.push(a//b)

    def _and(self):
        self.push(self.pop() & self.pop())

    def _or(self):
        self.push(self.pop() | self.pop())

    def _xor(self):
        self.push(self.pop() ^ self.pop())

    def fillOps(self):
        self.opcodes.append((self._rot, 5))
        self.opcodes.append((self._over, 5))
        self.opcodes.append((self._swap, 5))
        self.opcodes.append((self._div, 5))
        self.opcodes.append((self._or, 5))
        self.opcodes.append((self._and, 5))
        self.opcodes.append((self._equal, 5))
        self.opcodes.append((self._neg, 5))

        self.opcodes.append((self._plus, 4))
        self.opcodes.append((self._less, 4))
        self.opcodes.append((self._iftrue, 4))
        self.opcodes.append((self._xor, 4))
        self.opcodes.append((self._jmp, 4))
        self.opcodes.append((self._mul, 4))
        self.opcodes.append((self._out, 4))
        self.opcodes.append((self._dup, 4))

        for i in xrange(self.maxbits - 2):
            self.opcodes.append((self._push(i), 3 + i))

File: stack_machine.cc
#include "stack_machine.h"

StackMachine::StackMachine(int stack_size, long long default_stack_value, int max_jumps, int max_bits, int max_useful_outputs)
  : Machine(max_bits),
    stack_size_(stack_size),
    default_stack_value_(default_stack_value),
    max_jumps_(max_jumps),
    max_useful_outputs_(max_useful_outputs),
    current_nr_ops_(0){
  AddOps();
  stack_ = new long long[stack_size_];
}

StackMachine::~StackMachine(){
  delete[] stack_;
}

long long StackMachine::GetBits(int code_idx){
  if(current_nr_ops_ == 0){
    return code_idx;
  }

  if(code_idx >= kNrOps){
    return kConstPrefixMask | (1 << (code_idx - kNrOps + kConstPrefixLength));
  }
  return op_bits_[code_idx];
}

int StackMachine::ReadBits(long long code){
  if(current_nr_ops_ == 0){
    AddCode(code & 3);
    return 2;
  }

  if((code & kConstPrefixMask) == kConstPrefixMask){
    int i = kConstPrefixLength;
    while(!((code >> i) & 1)){
      i++;
      if(i == 40){
        printf("Something odd!\n");
        return 0;
      }
        
    }
    //printf("Const %d\n", i - kConstPrefixLength);
    AddCode(kNrOps + i - kConstPrefixLength);
    return i + 1;
  }

  /*NOTE: Can be made faster with lookup table*/
  for(int code_idx = 0; code_idx < kNrOps; code_idx++){
    int mask = (1 << nr_op_bits_[code_idx]) - 1;
    if((code & mask) == op_bits_[code_idx]){
      AddCode(code_idx);
      return nr_op_bits_[code_idx];
    }
  }
  return 0;
}

int StackMachine::GetNrBits(int code_idx){
  if(current_nr_ops_ == 0){
    return 2;
  }

  if(code_idx >= kNrOps){
    return kConstPrefixLength + 1 + code_idx - kNrOps;
  }
  return nr_op_bits_[code_idx];
}

int StackMachine::AddCode(int code_idx){
  int n = Machine::AddCode(code_idx);
  if(n){
    current_program_[current_nr_ops_++] = code_idx;
  }

  return n;
}

void StackMachine::DelCode(int nr_bits){
  current_nr_ops_--;
  Machine::DelCode(nr_bits);
}

void StackMachine::Load(long long code, int nr_bits){
  current_nr_ops_ = 0;
  Machine::Load(code, nr_bits);
}

int StackMachine::AddOp(std::string name, int bit_length, int next_ok){
  op_names_[current_added_ops_] = name;
  nr_op_bits_[current_added_ops_] = bit_length;

  //Reverse 8 bit integer
  int rbits = (next_ok * 0x0202020202ULL & 0x010884422010ULL) % 1023;

  op_bits_[current_added_ops_++] = rbits;

  next_ok += (1 << 7) >> (bit_length - 1);

  return next_ok;
}

void StackMachine::push(long long val){
  stackptr_ = (stackptr_ + 1) % stack_size_;
  stack_[stackptr_] = val;
}

long long StackMachine::peek(){
  return stack_[stackptr_];
}

long long StackMachine::pop(){
  long long val = stack_[stackptr_];
  stack_[stackptr_] = default_stack_value_;
  stackptr_--;
  if(stackptr_ == -1) stackptr_ = stack_size_ - 1;
  return val;
}

void StackMachine::Execute(){
  std::fill_n(stack_, stack_size_, default_stack_value_);

  int iptr = 0;
  int used_jumps = 0;
  bool cflag = false;

  int totinstr = 0;
  stackptr_ = 0;
  output_length_ = 0;
  long long temp1, temp2;

  AddSuffix(current_program_[0]);

  while(true){
    int op = current_program_[1 + (iptr++)];

    switch(op){
    case 0: //Add
      push(pop() + pop());
      break;
    case 1: //Mul
      push(pop() * pop());
      break;
    case 2: //XOR
      push(pop() ^ pop());
      break;
    case 3: //Less
      temp1 = pop(); //C++ does not guarantee evaluation order
      cflag = temp1 < pop();
      break;
    case 4: //If
      if(not cflag) iptr++;
      break;
    case 5: //Jmp
      if(used_jumps++ < max_jumps_) iptr = pop() % (current_nr_ops_ - 1);
      break;
    case 6: //Out
      output_[output_length_++] = peek();
      if(output_length_ == max_useful_outputs_) iptr = current_nr_ops_;
      break;
    case 7: //Dup
      push(peek());
      break;
    case 8: //Mod
      temp1 = pop();
      temp2 = pop();
      if(temp2 == 0) push(0);
      else push(temp1 % temp2);
      break;
    case 9: //Over
      temp1 = pop();
      temp2 = pop();
      push(temp1);
      push(temp2);
      push(temp1);
      break;
    case 10: //Swap
      temp1 = pop();
      temp2 = pop();
      push(temp2);
      push(temp1);
      break;
    case 11: //Div
      temp1 = pop();
      temp2 = pop();
      if(temp2 == 0) push(temp1);
      else push(temp1 / temp2);
      break;
    case 12: //OR
      push(pop() | pop());
      break;
    case 13: //AND
      push(pop() & pop());
      break;
    case 14: //Equal
      cflag = (pop() == pop());
      break;
    case 15: //Neg
      push(-pop());
      break;
    default: //Push const
      push(op - 16);
    }

    if(output_length_ == 0 && ++totinstr > kWakeup) break;
    

    if(iptr < 0) iptr = 0;
    else if(iptr >= current_nr_ops_) break;
  }
  current_nr_ops_ -= 3; //Remove suffix
}

int StackMachine::NrChoices(){
  if(nr_current_bits_ == 0){
    return 4;
  }

  int nrconsts = max_bits_ - (nr_current_bits_ + 2);
  if(nrconsts < (4 - 2)) return 0; //(smallest op - 2)
  return 16 + nrconsts;
}

void StackMachine::AddSuffix(int dest){
  current_program_[current_nr_ops_++] = 6;
  current_program_[current_nr_ops_++] = 16 + dest;
  current_program_[current_nr_ops_++] = 5;
}

void StackMachine::AddOps(){
  int next_ok = 0;
  current_added_ops_ = 0;
  next_ok = AddOp("Add", 4, next_ok);
  next_ok = AddOp("Mul", 4, next_ok);
  next_ok = AddOp("XOR", 4, next_ok);
  next_ok = AddOp("Less", 4, next_ok);
  next_ok = AddOp("If", 4, next_ok);
  next_ok = AddOp("Jmp", 4, next_ok);
  next_ok = AddOp("Out", 4, next_ok);
  next_ok = AddOp("Dup", 4, next_ok);

  next_ok = AddOp("Mod", 5, next_ok);
  next_ok = AddOp("Over", 5, next_ok);
  next_ok = AddOp("Swap", 5, next_ok);
  next_ok = AddOp("Div", 5, next_ok);
  next_ok = AddOp("OR", 5, next_ok);
  next_ok = AddOp("AND", 5, next_ok);
  next_ok = AddOp("Equal", 5, next_ok);
  AddOp("Neg", 5, next_ok);
}

std::string StackMachine::ShowCode(){
  std::string s = (boost::format("Setjmp %1%\n") % current_program_[0]).str();

  for(int i = 1; i < current_nr_ops_; i++){
    int op = current_program_[i];
    if(op >= kNrOps){
      s += (boost::format("Push %1%") % (op - kNrOps)).str();
    } else {
      s += op_names_[op];
    }
    s += "\n";
  }

  return s;
}

